README.md
# Web API Base SDK Bundle

Do you have an SDK library based off of our [web-api-base-sdk-php](https://github.com/northamericanbancard/web-api-base-sdk-php) 
package and wish to turn it into a Symfony service?

This 3rd Party Bundle will allow you to quickly create one or more `SDK`s as a service. You can choose to supply
either AWS IAM Credentials, a JWT token, or if there is no authentication required, no credentials for a `Simple SDK`.

If you have a library of one or more SDKs, where each SDK talks to a single API, and each SDK takes as
a single dependency a base client from [web-api-base-sdk-php](https://github.com/northamericanbancard/web-api-base-sdk-php) -
by creating your own Symfony 3rd Party Bundle and requiring this bundle and using the instructions below:
[Specialized Bundle](#specialized-bundle), you will be able to quickly turn your SDKs into Symfony Services with ease.

*_NOTE_*:

Even though the name of this bundle is the `web-api-base-sdk-bundle`, treat the word `base` as a bit of a misnomer. If you
just need a quick SDK to call out to endpoints without creating a specialized SDK bundle backed by this bundle, you are certainly
free to do so! Just follow the instructions under [Non-specialized Bundle](#non-specialized-bundle)

## Getting Started

Follow the below instructions in order to successfully include and utilize this bundle within your project.

### Prerequisites

Ensure you are running PHP >= 5.5 and Symfony >= 2.8. Any endpoints you are calling that are secured
via IAM, you will have to obtain the IAM user's access key and secret key. If required, you should
also obtain the `x-api-key` header's API key (found in API Gateway).  Any endpoints secured using JWT, will
require the JWT to be obtained as well.

The `x-api-key` header can be used with any SDK you create, giving you a consistent way to pass across an API key, regardless
of your authentication scheme - especially as your `Authorization: Bearer` header is already used for JWT.

## Installing

### Symfony

*_NOTE_*:

With either bundle architecture you use, you have the option to have your endpoint use JWT, AWS (IAM), 
or nothing as your primary authentication mechanism. Each option allows you to use an x-api-key header if you wish, 
for sending api keys (not generally used as anything more than to determine who the user is and track account details). 
JWT-based SDKs send requests with the header `Authorization: Bearer my.tok.en`

#### Non-specialized Bundle

1. Include this bundle in your Symfony project by running `composer require nab/web-api-base-sdk-bundle`.
2. Register the bundle with Symfony by editing your app/AppKernel.php: `new \NAB\Bundle\NAB\Bundle\WebApiBaseSdk\NABWebApiBaseSdkBundle(),`
3. update your app/config/config_* files with the proper configuration as noted below:

#### AWS

```yaml
nab_web_api_base_sdk:
  guzzle_configuration:
    http_errors: <DEFAULT:false>
    decode_content: <DEFAULT:true>
    verify: <DEFAULT:true>
    cookies: <DEFAULT:false>
  endpoints:
    my_endpoint_name:
      base_endpoint: <REQUIRED> #base path of your api such has https://my.api.aws.com/api/stage, no trailing /
      api_key: <DEFAULT:NULL>
      aws:
        aws_region: <DEFAULT:us-east-1>
        aws_service: <DEFAULT:execute-api>
        credentials:
          access_key: <IAM_USER_ACCESS_KEY_ID:REQUIRED>
          secret_key: <IAM_USER_SECRET_KEY:REQUIRED>
```

#### JWT

```yaml
nab_web_api_base_sdk:
  guzzle_configuration:
    http_errors: <DEFAULT:false>
    decode_content: <DEFAULT:true>
    verify: <DEFAULT:true>
    cookies: <DEFAULT:false>
  endpoints:
    my_endpoint_name:
      base_endpoint: <REQUIRED> #base path of your api such has https://my.api.aws.com/api/stage, no trailing /
      api_key: <DEFAULT:NULL>
      jwt:
        token: <MY_TOKEN:REQUIRED>
```

#### Simple

```yaml
nab_web_api_base_sdk:
  guzzle_configuration:
    http_errors: <DEFAULT:false>
    decode_content: <DEFAULT:true>
    verify: <DEFAULT:true>
    cookies: <DEFAULT:false>
  endpoints:
    my_endpoint_name:
      base_endpoint: <REQUIRED> #base path of your api such has https://my.api.aws.com/api/stage, no trailing /
      api_key: <DEFAULT:NULL>
      #Note - No authentication mechanism listed (such as jwt or aws).
```

Obtain the service using the service key `nab.web_api_sdk.<my_endpoint_name>.[aws|jwt|simple]_client` (depending on authentication scheme), and make a request
using the one of the following API calls:

#### API Calls:

The allowed calls you can make on the service are noted as:

1. httpGet(string $path, array $queryParams = [], array $headers = [], $body = null)
2. httpPost(string $path, array $queryParams = [], array $headers = [], $body = null)

*_NOTE_*: If your Guzzle configuration allows http_errors (true), the the above functions will throw
exceptions if you do not get 200 response codes. Review Guzzle's documentation regarding exceptions.

Example Calls:

```php
// Note the first '/' in the path (argument 1).
$service->httpGet('/foop/1', ['foo' => 'bar', 'baz' => 'bing'], ['Accept' => 'application/json'])
$service->httpPost('/bap', ['foo' => 'bar', 'baz' => 'bing'], ['Accept' => 'application/json'], '{"a": "b"}')
```

**_NOTE_** The client service is an extension of `GuzzleHttp\Client`, and therefore returns an instance of
`\Psr\Http\Message\ResponseInterface`, or throws an exception on failed requests.

#### Specialized Bundle

When including this bundle into your specialized bundle, do NOT register this bundle in your AppKernel. Instead,
inherit from our `DependencyInjection/(Configuration)&(NABWebApiBaseSdkExtension)`. One you've done that in your
config, override the extension's const `CONFIGURATION_ALIAS` variable in your extension class to be the alias you need 
for your bundle's configuration. Make the following addition to your `Configuration` class:

```php
// DependencyInjection\Configuration
/**
 * This method is used to get the root key for the configuration.
 *
 * {@see NABWebApiBaseSdkExtension::getAlias}
 *
 * @return string
 */
protected function getConfigurationRootKey()
{
    return (new MyBundleExtension())->getAlias();
}
```

Now, in order for Symfony to process your config's root key, add the following to your Extension file:

```php
/**
 * Overrides the parent configuration instance.
 *
 * @return Configuration
 */
protected function getConfigurationInstance()
    {
        return new Configuration();
    }
```

Now, add your endpoints as you would using your own bundles' config alias. The SDK classes that should wrap the
NAB Web Api Base SDK library's clients (such as JwtClient, AwsApiGatewayClient...etc), are specified using the config for `wrapped_by`.
If you omit the `getConfigurationInstance` from above, you will not have your bundle's config key, and it will instead be
`nab_web_api_base_sdk`.

If you need to add extra config options to what already exists, you can append them using the following in your
`Configuration` class:

```php
public function getConfigTreeBuilder()
    {
        $treeBuilder = parent::getConfigTreeBuilder();

        //protected attribute access workaround
        $reflectedClass = new \ReflectionObject($treeBuilder);
        $property = $reflectedClass->getProperty("root");
        $property->setAccessible(true);

        // Grab the rootnode from the actual existing TreeBuilder instance.
        $rootNode = $property->getValue($treeBuilder);

        // Append your config here.
        $rootNode
           ->children()
           ...

        return $treeBuilder;
     }
```

#### AWS

```yaml
my_special_config_alias:
  guzzle_configuration:
    http_errors: <DEFAULT:false>
    decode_content: <DEFAULT:true>
    verify: <DEFAULT:true>
    cookies: <DEFAULT:false>
  endpoints:
    my_endpoint_name:
      wrapped_by: My\Sdk\Class  #Either set this as YOUR SDK wrapper around the BaseSDK Client in your bundle by default during config load in your extension, or here, or leave empty.
      base_endpoint: <REQUIRED> #base path of your api such has https://my.api.aws.com/api/stage, no trailing /
      api_key: <DEFAULT:NULL>
      aws:
        aws_region: <DEFAULT:us-east-1>
        aws_service: <DEFAULT:execute-api>
        credentials:
          access_key: <IAM_USER_ACCESS_KEY_ID:REQUIRED>
          secret_key: <IAM_USER_SECRET_KEY:REQUIRED>
```

#### JWT

```yaml
nab_web_api_base_sdk:
  guzzle_configuration:
    http_errors: <DEFAULT:false>
    decode_content: <DEFAULT:true>
    verify: <DEFAULT:true>
    cookies: <DEFAULT:false>
  endpoints:
    my_endpoint_name:
      wrapped_by: My\Sdk\Class  #Either set this as YOUR SDK wrapper around the BaseSDK Client in your bundle by default during config load in your extension, or here, or leave empty.
      base_endpoint: <REQUIRED> #base path of your api such has https://my.api.aws.com/api/stage, no trailing /
      api_key: <DEFAULT:NULL>
      jwt:
        token: <MY_TOKEN:REQUIRED>
```

#### Basic

```yaml
nab_web_api_base_sdk:
  guzzle_configuration:
    http_errors: <DEFAULT:false>
    decode_content: <DEFAULT:true>
    verify: <DEFAULT:true>
    cookies: <DEFAULT:false>
  endpoints:
    my_endpoint_name:
      wrapped_by: My\Sdk\Class  #Either set this as YOUR SDK wrapper around the BaseSDK Client in your bundle by default during config load in your extension, or here, or leave empty.
      base_endpoint: <REQUIRED> #base path of your api such has https://my.api.aws.com/api/stage, no trailing /
      api_key: <DEFAULT:NULL>
      basic:
        username: <MY_HTTP_BASIC_AUTH_USERNAME:REQUIRED>
        password: <MY_HTTP_BASIC_AUTH_PASSWORD:REQUIRED>
```

#### Simple

```yaml
nab_web_api_base_sdk:
  guzzle_configuration:
    http_errors: <DEFAULT:false>
    decode_content: <DEFAULT:true>
    verify: <DEFAULT:true>
    cookies: <DEFAULT:false>
  endpoints:
    my_endpoint_name:
      wrapped_by: My\Sdk\Class  #Either set this as YOUR SDK wrapper around the BaseSDK Client in your bundle by default during config load in your extension, or here, or leave empty.
      base_endpoint: <REQUIRED> #base path of your api such has https://my.api.aws.com/api/stage, no trailing /
      api_key: <DEFAULT:NULL>
      #Note - No authentication mechanism listed (such as jwt or aws).
```

#### API Calls

This is entirely dependent on how you decided to wrap the calls found under the Non-Specialized version.

## Development

**_First Install/Make Introduction_**

1.  Ensure you have autoconf (v2.69 at min) installed.
2.  In the project root, run `make` to install your new project.
3.  View the Makefile, or Makefile.in for further commands and required inputs.
4.  Any edits to Makefile should be done in Makefile.in and then run `autoconf && ./configure && make`

## Testing and Maintenance

The following command should be run often, and should be run during every code review:

```bash
make test
```

OR

```bash
make docker-test
```

OR

```bash
make php
```

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/northamericanbancard/web-api-base-sdk-bundle/tags). 

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.4.0] - 2018-05-08
### Added
- Added in the ability to set, via guzzle_config bundle config key, the options `allow_redirects (default true)`, `cert (default null)`,
`connect_timeout (default 5.0 seconds)`, `debug (default false)`, `headers (default [])`, `ssl_key (default null)`, `stream (default false)`, and
`timeout (default 5.0 seconds)`.


Tests/DependencyInjection/ConfigurationTest.php
<?php

/**
 * Filename:    ConfigurationTest.php
 * Created:     14/02/18, at 6:00 PM
 * @author      James Hollenbeck <jhollenbeck@nabancard.com>
 * @copyright   1992-2018 North American Bancard
 */

namespace NAB\Bundle\WebApiBaseSdk\Tests\DependencyInjection;

use NAB\Bundle\WebApiBaseSdk\DependencyInjection\Configuration;
use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;

class ConfigurationTest extends KernelTestCase
{
    /**
     * @dataProvider configurationCanControlGuzzleConfigDataProvider
     *
     * @param array $guzzleConfigSupplied
     * @param array $expected
     */
    public function testConfigurationCanControlGuzzleConfig(array $guzzleConfigSupplied, array $expected)
    {
        $treeBuilder = (new Configuration())->getConfigTreeBuilder();
        $tree = $treeBuilder->buildTree();
        $finalConfig = $tree->finalize($tree->normalize($guzzleConfigSupplied));
        $finalConfig = $finalConfig['guzzle_configuration'];

        ksort($expected);
        ksort($finalConfig);

        $this->assertSame($expected, $finalConfig);
    }

    /**
     * @return array
     */
    public function configurationCanControlGuzzleConfigDataProvider()
    {
        return [
            [
                'input' => [
                    'guzzle_configuration' => []
                ],
                'expected' => [
                    'http_errors' => false,
                    'decode_content' => true,
                    'verify' => true,
                    'cookies' => false,
                    'allow_redirects' => true,
                    'cert' => null,
                    'connect_timeout' => 5.0,
                    'debug' => false,
                    'headers' => [],
                    'ssl_key' => null,
                    'stream' => false,
                    'timeout' => 5.0,
                ]
            ],
            [
                'input' => [
                    'guzzle_configuration' => [
                        'http_errors' => true,
                        'cookies' => true,
                        'timeout' => 6.0,
                    ]
                ],
                'expected' => [
                    'http_errors' => true,
                    'decode_content' => true,
                    'verify' => true,
                    'cookies' => true,
                    'allow_redirects' => true,
                    'cert' => null,
                    'connect_timeout' => 5.0,
                    'debug' => false,
                    'headers' => [],
                    'ssl_key' => null,
                    'stream' => false,
                    'timeout' => 6.0,
                ]
            ]
        ];
    }
}


Tests/DependencyInjection/NABWebApiBaseSdkExtensionTest.php
<?php

/**
 * Filename:    NABWebApiBaseSdkExtensionTest.php
 * Created:     16/10/17, at 2:26 PM
 * @author      James Hollenbeck <jhollenbeck@nabancard.com>
 * @copyright   1992-2017 North American Bancard
 */

namespace NAB\Bundle\WebApiBaseSdk\Tests\DependencyInjection;

use Aws\Credentials\Credentials;
use NAB\Bundle\WebApiBaseSdk\DependencyInjection\Configuration;
use NAB\WebApiBaseSdk\AwsApiGatewayClient;
use NAB\Bundle\WebApiBaseSdk\DependencyInjection\NABWebApiBaseSdkExtension;
use NAB\WebApiBaseSdk\BasicClient;
use NAB\WebApiBaseSdk\JwtClient;
use NAB\WebApiBaseSdk\SignatureV4;
use NAB\WebApiBaseSdk\SimpleClient;
use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;
use Symfony\Component\DependencyInjection\ContainerBuilder;
use Symfony\Component\DependencyInjection\Definition;

/**
 * Test the various ways we get services created, and that they 'seem' correct in terms of constructor
 * arguments and service definition class names.
 */
class NABWebApiBaseSdkExtensionTest extends KernelTestCase
{
    /**
     * @var NABWebApiBaseSdkExtension
     */
    private $systemUnderTest;

    /**
     * {@inheritdoc}
     */
    protected function setUp()
    {
        $this->systemUnderTest = new NABWebApiBaseSdkExtension();
    }

    /**
     * {@inheritdoc}
     */
    protected function tearDown()
    {
        unset($this->systemUnderTest);
    }

    public function extensionWillSetExpectedServiceDataProvider()
    {
        $testDataBase = [
            'nab_web_api_base_sdk' => [
                'guzzle_configuration' => [],
                'endpoints' => [
                    'test' => [
                        'base_endpoint' => 'test.com',
                        'api_key' => 'abc',
                    ],
                    'derp' => [
                        'base_endpoint' => 'derp.com',
                        'api_key' => 'abc',
                    ]
                ]
            ]
        ];

        $credentials = [
            'jwt' => ['jwt' => ['token' => 'my.tok.en']],
            'aws' => ['aws' => ['credentials' => ['access_key' => 1, 'secret_key' => 2,]]],
            'basic' => ['basic' => ['username' => 'user', 'password' => 'password',]],
            'simple' => [],
        ];

        $guzzleConfigurationBase = [
            'http_errors' => false,
            'decode_content' => true,
            'verify' => true,
            'cookies' => false,
            'allow_redirects' => true,
            'cert' => null,
            'connect_timeout' => 5.0,
            'debug' => false,
            'headers' => [],
            'ssl_key' => null,
            'stream' => false,
            'timeout' => 5.0,
        ];

        $jwtConfiguration = $testDataBase;
        $awsConfiguration = $testDataBase;
        $simpleConfiguration = $testDataBase;
        $basicConfiguration = $testDataBase;

        foreach (['test', 'derp'] as $clientKey) {
            foreach ([
                'jwt' => &$jwtConfiguration,
                'aws' => &$awsConfiguration,
                'simple' => &$simpleConfiguration,
                'basic' => &$basicConfiguration,
            ] as $key => &$testData) {
                    $testData['nab_web_api_base_sdk']['endpoints'][$clientKey] += $credentials[$key];
            }
        }

        return [
            'simple' => [
                'configuration' => $simpleConfiguration,
                'class_name' => SimpleClient::class,
                'service_constructor_argument_length' => 3,
                'constructor_definitions' => [],
                'x_api_token_argument_index' => 1,
                'guzzle_config_argument_index' => 2,
                'expected_guzzle_config' => $guzzleConfigurationBase,
                'service_classification' => 'simple',
            ],
            'basic' => [
                'configuration' => $basicConfiguration,
                'class_name' => BasicClient::class,
                'service_constructor_argument_length' => 5,
                'constructor_definitions' => [],
                'x_api_token_argument_index' => 3,
                'guzzle_config_argument_index' => 4,
                'expected_guzzle_config' => $guzzleConfigurationBase,
                'service_classification' => 'basic',
            ],
            'jwt' => [
                'configuration' => $jwtConfiguration,
                'class_name' => JwtClient::class,
                'service_constructor_argument_length' => 4,
                'constructor_definitions' => [],
                'x_api_token_argument_index' => 2,
                'guzzle_config_argument_index' => 3,
                'expected_guzzle_config' => $guzzleConfigurationBase,
                'service_classification' => 'jwt',
            ],
            'aws' => [
                'configuration' => $awsConfiguration,
                'class_name' => AwsApiGatewayClient::class,
                'service_constructor_argument_length' => 5,
                'constructor_definitions' => [
                    1 => SignatureV4::class,
                    2 => Credentials::class,
                ],
                'x_api_token_argument_index' => 3,
                'guzzle_config_argument_index' => 4,
                'expected_guzzle_config' => $guzzleConfigurationBase,
                'service_classification' => 'aws',
            ],
        ];
    }

    /**
     * I know this looks like a wonky, integration type test, but because the ContainerBuilder is really
     * a POPO with some factor magic - we really don't need to worry about mocking it, and can rely of tracking
     * its Definition classes (the proxies that exist before an actual service is created).
     *
     * Sincerely,
     *  - mgmt
     *
     * @dataProvider extensionWillSetExpectedServiceDataProvider
     *
     * @param array  $configuration                         The config to pass into our Extension class
     * @param string $expectedClassName                     The class-name of the service to be built
     * @param int    $expectedConstructorArgumentListLength The number of arguments expected to be passed to our class
     * @param array  $expectedConstructorDefinitions        For any arguments that are Definitions, verify class names
     * @param int    $expectedXApiKeyArgumentIndex          Which constructor argument index is the API key
     * @param int    $expectedGuzzleConfigArgumentIndex     Which constructor argument index is the guzzle config
     * @param array  $expectedGuzzleConfig                  What we expect the guzzle config to be upon instantiation
     * @param string $serviceClassification                 A key [simple, jwt, aws] that classifies the client
     */
    public function testExtensionWillSetExpectedService(
        array $configuration,
        $expectedClassName,
        $expectedConstructorArgumentListLength,
        array $expectedConstructorDefinitions,
        $expectedXApiKeyArgumentIndex,
        $expectedGuzzleConfigArgumentIndex,
        array $expectedGuzzleConfig,
        $serviceClassification
    ) {
        $containerBuilder = new ContainerBuilder();

        $this->systemUnderTest->load(
            $configuration,
            $containerBuilder
        );

        $clients = ['test', 'derp'];
        $this->assertCount(count($clients), $containerBuilder->getDefinitions());

        foreach ($clients as $client) {
            $serviceKey = sprintf('nab.web_api_sdk.%s.%s_client', $client, $serviceClassification);
            // Assert we still have a service definition.
            $definition = $containerBuilder->getDefinition($serviceKey);
            $this->assertInstanceOf(
                'Symfony\Component\DependencyInjection\Definition',
                $definition,
                'Failed to assert service is instance of Definition.'
            );

            $this->assertSame($expectedClassName, $definition->getClass());

            // Assert that the service definition was created with 3 arguments.
            $this->assertCount(
                $expectedConstructorArgumentListLength,
                $definition->getArguments(),
                'Failed to assert the expected number of constructor arguments.'
            );

            // Assert, if expected, that other Definition classes were created for construct.
            if ($expectedConstructorDefinitions) {
                foreach ($expectedConstructorDefinitions as $index => $className) {
                    /** @var Definition */
                    $argumentDefinition = $definition->getArguments()[$index];
                    $this->assertSame(
                        $argumentDefinition->getClass(),
                        $className,
                        'Failed asserting other constructor args are Definitions for ' . $serviceClassification . '.'
                    );
                }
            }

            // Almost there, assert the api key was set
            $this->assertSame(
                'abc',
                $definition->getArgument($expectedXApiKeyArgumentIndex),
                'Failed asserting api key was set as expected.'
            );

            // Lastly,
            $guzzleConfiguration = $definition->getArgument($expectedGuzzleConfigArgumentIndex);

            ksort($expectedGuzzleConfig);
            ksort($guzzleConfiguration);

            $this->assertSame(
                $expectedGuzzleConfig,
                $guzzleConfiguration,
                'Failed asserting guzzle config was set as expected.'
            );
        }
    }

    public function testExtensionIsDefiningExpectedAlias()
    {
        $expected = 'nab_web_api_base_sdk';
        $this->assertSame($expected, $this->systemUnderTest->getAlias());
    }

    public function testBaseSdkCanBeWrapped()
    {
        $configuration = [
            'nab_web_api_base_sdk' => [
                'guzzle_configuration' => [],
                'endpoints' => [
                    'test' => [
                        'wrapped_by' => NABWebApiBaseSdkExtension::class,
                        'base_endpoint' => 'test.com',
                        'api_key' => 'abc',
                    ],
                    'derp' => [
                        'wrapped_by' => Configuration::class,
                        'base_endpoint' => 'derp.com',
                        'api_key' => 'abc',
                    ]
                ]
            ]
        ];

        $containerBuilder = new ContainerBuilder();

        $this->systemUnderTest->load(
            $configuration,
            $containerBuilder
        );

        foreach (['test', 'derp'] as $endpoint) {
            $definitionKey = sprintf('nab.web_api_sdk.%s.simple_client', $endpoint);
            $this->assertTrue(
                $containerBuilder->hasDefinition($definitionKey)
            );

            $definition = $containerBuilder->getDefinition($definitionKey);
            $this->assertSame(
                $configuration['nab_web_api_base_sdk']['endpoints'][$endpoint]['wrapped_by'],
                $definition->getClass()
            );
        }
    }

    public function testBaseSdkIsNotWrappedByDefault()
    {
        $configuration = [
            'nab_web_api_base_sdk' => [
                'guzzle_configuration' => [],
                'endpoints' => [
                    'test' => [
                        'base_endpoint' => 'test.com',
                        'api_key' => 'abc',
                    ],
                    'derp' => [
                        'base_endpoint' => 'derp.com',
                        'api_key' => 'abc',
                    ]
                ]
            ]
        ];

        $containerBuilder = new ContainerBuilder();

        $this->systemUnderTest->load(
            $configuration,
            $containerBuilder
        );

        foreach (['test', 'derp'] as $endpoint) {
            $definitionKey = sprintf('nab.web_api_sdk.%s.simple_client', $endpoint);
            $this->assertTrue(
                $containerBuilder->hasDefinition($definitionKey)
            );

            $definition = $containerBuilder->getDefinition($definitionKey);
            $this->assertSame(
                SimpleClient::class,
                $definition->getClass()
            );
        }
    }
}


DependencyInjection/Configuration.php
<?php

/**
 * Filename:    Configuration.php
 * Created:     13/10/17, at 3:07 PM
 * @author      James Hollenbeck <jhollenbeck@nabancard.com>
 * @copyright   1992-2017 North American Bancard
 */

namespace NAB\Bundle\WebApiBaseSdk\DependencyInjection;

use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

/**
 * Configuration class that sets up the what this bundle expects in terms of
 * config*.yml files in main projects.
 */
class Configuration implements ConfigurationInterface
{
    /**
     * {@inheritdoc}
     */
    public function getConfigTreeBuilder()
    {
        $treeBuilder = new TreeBuilder();
        $rootNode = $treeBuilder->root($this->getConfigurationRootKey());

        $rootNode
            ->children()
                ->arrayNode('guzzle_configuration')
                    ->info('Set the configuration as expected by a GuzzleHttp\Client')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('http_errors')->defaultFalse()->end()
                        ->scalarNode('decode_content')->defaultTrue()->end()
                        ->scalarNode('verify')->defaultTrue()->end()
                        ->scalarNode('cookies')->defaultFalse()->end()
                        ->scalarNode('allow_redirects')->defaultTrue()->end()
                        ->scalarNode('cert')->defaultNull()->end()
                        ->scalarNode('connect_timeout')->defaultValue(5.0)->end()
                        ->scalarNode('debug')->defaultFalse()->end()
                        ->scalarNode('headers')->defaultValue([])->end()
                        ->scalarNode('ssl_key')->defaultNull()->end()
                        ->scalarNode('stream')->defaultFalse()->end()
                        ->scalarNode('timeout')->defaultValue(5.0)->end()
                    ->end()
                ->end()
                ->arrayNode('endpoints')
                ->info('The collection of base-urls and SDK configuration')
                ->useAttributeAsKey('name')
                ->requiresAtLeastOneElement()
                ->prototype('array')
                    ->children()
                        ->scalarNode('wrapped_by')
                            ->info(
                                'If this bundle is being used to turn a library SDK into a service, map each endpoint' .
                                ' to the proper library SDK class. This SDK lib should be based off of the' .
                                ' web-api-base-sdk lib.'
                            )
                            ->defaultNull()
                            ->end()
                        ->scalarNode('base_endpoint')->info('The base URL, no trailing /')->isRequired()->end()
                        ->scalarNode('api_key')
                            ->info('The x-api-header value (usually for AWS APIG)')
                            ->defaultNull()
                        ->end()
                        ->arrayNode('jwt')
                            ->info('Once enabled, use JWT as your authentication service: "Authorization: Bearer".')
                            ->canBeEnabled()
                            ->children()
                                ->scalarNode('token')->isRequired()->end()
                            ->end()
                        ->end()
                        ->arrayNode('basic')
                            ->info('Once enabled, use Basic as your authentication service: "Authorization: Basic".')
                            ->canBeEnabled()
                            ->children()
                                ->scalarNode('username')->isRequired()->end()
                                ->scalarNode('password')->isRequired()->end()
                            ->end()
                        ->end()
                        ->arrayNode('aws')
                            ->info('Once enabled, use IAM Access/Secrete keys for authentication.')
                            ->canBeEnabled()
                            ->children()
                                ->scalarNode('aws_region')->defaultValue('us-east-1')->end()
                                ->scalarNode('aws_service')->defaultValue('execute-api')->end()
                                ->arrayNode('credentials')
                                    ->isRequired()
                                    ->children()
                                        ->scalarNode('access_key')->isRequired()->end()
                                        ->scalarNode('secret_key')->isRequired()->end()
                                    ->end()
                                ->end()
                            ->end()
                        ->end()
                    ->end()
                ->end()
            ->end()
        ;

        return $treeBuilder;
    }

    /**
     * This method is used to get the root key for the configuration. If you are creating a specialized bundle,
     * inherit from this class, and override this function to return (new MyConfiguration())->getAlias().
     *
     * {@see NABWebApiBaseSdkExtension::getAlias}
     *
     * @return string
     */
    protected function getConfigurationRootKey()
    {
        return (new NABWebApiBaseSdkExtension())->getAlias();
    }
}
